### ISSUE TITLE
Make CatalogTable columns and row data configurable

### ISSUE THREAD

**Issue by heatheralee** [backstage/backstage#5536]
## Feature Suggestion

Hi there! We would like to customize the home page table, `CatalogTable` with our own columns. It appears that right now the [columns are hardcoded to a specific set of columns](https://github.com/backstage/backstage/blob/master/plugins/catalog/src/components/CatalogTable/CatalogTable.tsx#L6) and there's not any way to configure these columns at this moment. In order to update the UI in order to overwrite certain core plugins, we have been using [`npm patch-package`](https://www.npmjs.com/package/patch-package), but I am afraid as time goes on this approach will become increasingly unwieldy.

## Possible Implementation

There's several ways to approach this:
1. Update `CatalogPage.tsx` to take in an additional properties: `columns` and `rows`, which are passed down to the `<CatalogTable>` component. If they are not defined, default to the current behavior. If they are defined, overwrite the existing columns and row data with what is passed in to the prop.
2. Replace [`<CatalogTable ... />`](https://github.com/backstage/backstage/blob/5c831724ec49ea1c76cfa6a7f6d130a6ecf5909b/plugins/catalog/src/components/CatalogPage/CatalogPage.tsx#L210) with `{children}` and allow `CatalogPage` to have a nested component which renders your own version of `CatalogTable`.

## Context

We are currently trying to update the home page table so that the columns are configurable on an application basis.

**Comment by timbonicus** [comment]
Yes, this is definitely needed! There's an existing issue #5070 that may cover the same suggestion.

**Comment by freben** [comment]
See my comment in the linked ticket above.

@GoWind talked about doing this in a piecemeal fashion. How far could we come by just adding a `columns` prop to `CatalogTable`, which gets passed down from the page?

See how the EntityTable handles columns, including exposing the defaults etc, making it pretty nicely composable https://github.com/backstage/backstage/blob/master/plugins/catalog-react/src/components/EntityTable/EntityTable.tsx#L77

**Comment by GoWind** [comment]
@freben ,  custom columns prop would be a very good start. 
Let me check with my team to see if we need more customization, possible being able to pass the CatalogTable as a prop.

**Comment by yousifalraheem** [comment]
Great suggestion .. The best solution is to convert Backstage UI Components into compositions. Allowing us to compose the UI while customizing all of its different parts is the ultimate solution. Instead of this:

```jsx
<CatalogTable entities={entities} />
```
We could have this:

```jsx
<CatalogTable>
  <CatalogTableHeaders headers={headers} />
  <CatalogTableBody>
    {list.map(item => <CatalogTableCell {...item} />)}
  </CatalogTableBody>
<CatalogTable/>
```

But also make sure that each of the these components extends their native elements props. We should be able to pass props like `title`, `onClick`, `onMouseEnter`, `aria-label`, etc.

**Comment by freben** [comment]
Yeah don't pass components as props unless necessary. Composition is the primary pattern where it's convenient to do so.

**Comment by heatheralee** [comment]
Hi @freben, I just created a PR here to take in a `columns` prop: https://github.com/backstage/backstage/pull/5593 where I followed the structure of EntityTable in terms of defining `columns` -- is this what you had in mind? Thanks in advance!
